impl/client/src/main/javascript/web/dojo/pentaho/common/Dialog.js

```diff
@@ -14,316 +14,341 @@
+    },
+
+    /**
+     * Sets the element opening the dialog and to which focus should be restored when the dialog is hidden.
+     *
+     * Supports WCAG focus maintenance.
+     *
+     * @param {Element} dialogOpener - The element opening the dialog.
+     */
+    setDialogOpener: function(dialogOpener) {
+      this.dialogOpener = dialogOpener;
+    },
+
+    // ElementRefresher: function(Element) : Element
+    // Supports WCAG focus maintenance.
+    // Allows "refreshing" the element to use for "refocus" to a corresponding one which is attached
+    // to the document. Required to deal with UI elements which are discarded and rebuilt.
+    elementRefresher: function(elem) {
+      return focusUtil.refreshElement(elem);
+    },
+
+    setElementRefresher: function(refresher) {
+      if (refresher != null) {
+        this.elementRefresher = refresher;
+      } else {
+        delete this.elementRefresher;
+      }
+    },
+
+    setLocalizationLookupFunction: function(f) {
+      this.getLocaleString = f;
+      this._localize();
+    },
+
+    buttonClick: function(idx) {
+      if(this.callbacks[idx]) {
+        this.callbacks[idx]();
+      }
+    },
+
+    // override this..
+    buttons: [],
+
+    // override this..
+    _localize: function() {
+      if(this.getLocaleString) {
+        for(var i=0; i<this.buttons.length; i++) {
+          var button = query("button"+i, this.popup.domNode);
+          this.buttons[i] = this.getSanitizedLocaleString(this.buttons[i]);
+          if(button) {
+            button.innerHTML = this.buttons[i];
+          }
+        }
+        if (this.hasCloseIcon) {
+          this.closeIcon.domNode.setAttribute("title", this.getLocaleString('CloseIcon_title'));
+        }
+      }
+    },
+
+    // override this
+    templateString : "<div></div>",
+
+    // * set variables for the template
+    postMixInProperties: function() {
+      this.inherited(arguments);
+    },
+
+    _createButtonPanel: function() {
+      this.buttonPanel = construct.create("TABLE");
+      style.set(this.buttonPanel, "width", "100%");
+      construct.place(this.buttonPanel, this.popup.containerNode);
+      domClass.add(this.buttonPanel, 'button-panel');
+    },
+
+    _createButtons: function() {
+      if (this.buttonPanel.rows.length > 0) {
+        this.buttonPanel.deleteRow(-1);
+      }
+      var row = this.buttonPanel.insertRow(-1);
+      var cell = row.insertCell(-1);
+      style.set(cell, "align", "right");
+      style.set(cell, "width", "100%");
+      for(var j=0; j<this.buttons.length; j++) {
+        cell = row.insertCell(-1);
+        style.set(cell, "align", "right");
+        var btn = document.createElement("BUTTON");
+        domClass.add(btn, "pentaho-button");
+        btn.setAttribute( "id", "button"+j);
+        // this is filled with the sanitized localized strings, however buttons can be override hence the double check
+        btn.innerHTML = this.getSanitizedLocaleString(this.buttons[j]);
```

Comment:
Although I would tend to trust that `buttons` hasn't been tampered with, and that it is only set in the prototype, I might be best to be more cautious. However, in this case, I would only repeat the `DOMPurify.sanitize` part (directly calling it) and not the localization bundle lookup part. While only rarely, a key will resolve to something which is also an existing key, and that some rare cases out there may be calling `getLocaleString` twice, I think we should avoid doing that now, always and push our luck. What do you think?

Reply:
Valid point, change done.